framework/src/araui/assets/ui_font_asset.c

```diff
@@ -0,0 +1,280 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <string.h>
+#include <stdint.h>
+#include <sys/types.h>
+#include <araui/ui_commons.h>
+#include <araui/ui_asset.h>
+#include "libs/ui_map.h"
+#include "libs/ui_list.h"
+#include "ui_core_internal.h"
+#include "ui_asset_internal.h"
+#include "ui_commons_internal.h"
+#include "ui_request_callback.h"
+#include "ui_log.h"
+
+#define DEFAULT_GLYPH_MAP_CAPACITY 256
+
+static void _ui_font_asset_destroy_func(void *userdata);
+
+ui_asset_t ui_font_asset_create_from_file(const char *filename)
+{
+	ui_font_asset_body_t *body;
+	FILE *file;
+	uint32_t font_height;
+	uint32_t font_width;
+	uint32_t glyph_num;
+	uint32_t utf8_code;
+	uint32_t data_size;
+	int glyph_idx;
+	char *extension;
+	ui_asset_glyph_t *glyph;
+
+	if (!ui_is_running()) {
+		UI_LOGE("error: UI framework is not running!\n");
+		return UI_NULL;
+	}
+
+	if (!filename) {
+		UI_LOGE("error: invaild parameter!\n");
+		return UI_NULL;
+	}
+
+	extension = strrchr(filename, '.');
+	if (strncmp(extension, ".fnt", 5) != 0) {
+		UI_LOGE("error: unsupported font file!\n");
+		return UI_NULL;
+	}
+
+	// file open
+	file = fopen(filename, "r");
+	if (!file) {
+		UI_LOGE("error: failed to open file!\n");
+		return UI_NULL;
+	}
+
+	// read header of file
+	ui_fread(&font_height, 4, 1, file);
+	ui_fread(&glyph_num, 4, 1, file);
+
+	// make font asset body
+	body = (ui_font_asset_body_t *)UI_ALLOC(sizeof(ui_font_asset_body_t));
+	if (!body) {
+		(void)fclose(file);
+		UI_LOGE("error: out of memory!\n");
+		return UI_NULL;
+	}
+
+	memset(body, 0, sizeof(ui_font_asset_body_t));
+	((ui_asset_body_t *)body)->type = UI_FONT_ASSET;
+	body->height = font_height;
+
+	// read glyph from the buffer
+	ui_map_init(&body->glyph_map, DEFAULT_GLYPH_MAP_CAPACITY);
+	for (glyph_idx = 0; glyph_idx < glyph_num; glyph_idx++) {
+		ui_fread(&utf8_code, 4, 1, file);
+		ui_fread(&font_width, 4, 1, file);
+		ui_fread(&data_size, 4, 1, file);
+
+		glyph = (ui_asset_glyph_t *)UI_ALLOC(sizeof(ui_asset_glyph_t));
+		if (!glyph) {
+			_ui_font_asset_destroy_func((void *)body);
+			(void)fclose(file);
+			UI_LOGE("error: out of memory!\n");
+			return UI_NULL;
+		}
+
+		glyph->data = (uint8_t *)UI_ALLOC(data_size);
+		if (!glyph->data) {
+			UI_FREE(glyph);
+			_ui_font_asset_destroy_func((void *)body);
+			(void)fclose(file);
+			UI_LOGE("error: out of memory!\n");
+			return UI_NULL;
+		}
+
+		ui_fread(glyph->data, 1, data_size, file);
+
+		glyph->size = data_size;
+		glyph->width = font_width;
+
+		ui_map_push(&body->glyph_map, utf8_code, glyph);
```

Comment:
What if this function fails?
Should be considered.

Reply:
I'll add return type and the error logs.
This logic can skip the error (only the logging is needed).
Because unadded glyph is not showing.